std/algorithm/iteration.d

```diff
@@ -4280,3 +4284,263 @@ unittest
          [1, 2, 0],
          [2, 1, 0]]));
 }
+
+/**
+Finds the mean (colloquially known as the average) of a range.
+This function is $(D O(r.length)).
+
+Params:
+    r = an input range
+
+Returns:
+    the mean of r
+*/
+auto mean(R)(R r)
+if (isInputRange!R &&
+    !isInfinite!R &&
+    is(ElementType!R : double))
+in
+{
+    assert(!r.empty);
+}
+body
+{
+    static if (hasLength!R)
+    {
+        if (r.length == 0)
+        {
+            return double.nan;
+        }
+        else if (r.length == 1)
+        {
+            return r.front;
+        }
+
+        return sum(r) / r.length;
+    }
+    else
+    {
+        import std.typecons : tuple;
+
+        auto pair = reduce!((a, b) => tuple(a[0] + 1, a[1] + b))
+            (tuple(size_t(0), 0.0), r);
+
+
+        if (pair[0] == 0)
+        {
+            return double.nan;
+        }
+
+        return pair[1] / pair[0];
+    }
+}
+
+///
+@safe unittest
+{
+    import std.math : approxEqual;
+
+    assert(mean([1, 2, 3]) == 2);
+    assert(approxEqual(mean([1.5, 2.5, 12.5]), 5.5));
+}
+
+unittest
+{
+    import std.internal.test.dummyrange : ReferenceInputRange;
+    import std.math : approxEqual;
+
+    auto r1 = new ReferenceInputRange!int([1, 2, 3]);
+    assert(r1.mean == 2);
+
+    auto r2 = new ReferenceInputRange!double([1.5, 2.5, 12.5]);
+    assert(approxEqual(r2.mean, 5.5));
+}
+
+/**
+Function to compute the quantile of a $(D SortedRange) given $(D percentile).
+If the range is empty, this function will fail an assert in debug mode and
+return NaN in release mode. Is $(D O(r.length * percentile)).
+
+Params:
+    r = a forward range with length
+    percentile = where to calculate the quantile
+
+Returns:
+    the quantile of the range given $(D percentile)
+*/
+double quantile(alias less = "a < b", R)(SortedRange!(R, less) r, double percentile) if (
+    isForwardRange!R &&
+    hasLength!R &&
+    !isRandomAccessRange!R &&
+    !isInfinite!R &&
+    is(ElementType!R : double))
+in
+{
+    assert(!r.empty);
+}
+body
+{
+    import std.math : modf, approxEqual;
+
+    if (r.length == 0)
+    {
+        return double.nan;
+    }
+    else if (r.length == 1)
+    {
+        return r.front;
+    }
+
+    double x = r.length * percentile;
+    real j;
+    real g = modf(x, j);
+    uint j_int = cast(uint) j;
```

Comment:
the lines between 4383 to 4397 are equal between this functions and the next. maybe you can merge this to make it more DRY